OMCompiler/Compiler/SimCode/SerializeSparsityPattern.mo

```diff
@@ -0,0 +1,139 @@
+/*
+ * This file is part of OpenModelica.
+ *
+ * Copyright (c) 1998-CurrentYear, Open Source Modelica Consortium (OSMC),
+ * c/o Linköpings universitet, Department of Computer and Information Science,
+ * SE-58183 Linköping, Sweden.
+ *
+ * All rights reserved.
+ *
+ * THIS PROGRAM IS PROVIDED UNDER THE TERMS OF GPL VERSION 3 LICENSE OR
+ * THIS OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
+ * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
+ * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the Open Source Modelica
+ * Consortium (OSMC) Public License (OSMC-PL) are obtained
+ * from OSMC, either from the above address,
+ * from the URLs: http://www.ida.liu.se/projects/OpenModelica or
+ * http://www.openmodelica.org, and in the OpenModelica distribution.
+ * GNU version 3 is obtained from: http://www.gnu.org/copyleft/gpl.html.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without
+ * even the implied warranty of  MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE, EXCEPT AS EXPRESSLY SET FORTH
+ * IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE CONDITIONS OF OSMC-PL.
+ *
+ * See the full OSMC Public License conditions for more details.
+ *
+ */
+
+encapsulated package SerializeSparsityPattern
+
+import List;
+import SimCode;
+import Util;
+
+function serialize
+  input SimCode.SimCode code;
+  output String dummy = "";
+protected
+  array<Integer> columnPointers, rowIndices, columns;
+  String fname;
+algorithm
+  for jac in code.jacobianMatrices loop
+    if not listEmpty(jac.sparsity) then
+      fname := code.fileNamePrefix + "_Jac" + jac.matrixName + ".bin";
+      columnPointers := listArray(0 :: list(listLength(Util.tuple22(column)) for column in jac.sparsity));
+      rowIndices := listArray(List.flatten(list(Util.tuple22(column) for column in jac.sparsity)));
+      serializeJacobian(fname, arrayLength(columnPointers), arrayLength(rowIndices), columnPointers, rowIndices);
+      for color in jac.coloredCols loop
+        columns := listArray(color);
+        serializeColor(fname, arrayLength(columns), columns);
+      end for;
+    end if;
+  end for;
+end serialize;
+
+// *********************
+// write to binary stuff
+// *********************
+
+protected
+  function serializeJacobian
+    input String name;
+    input Integer numCols;
+    input Integer nnz;
+    input array<Integer> colPtrs;
+    input array<Integer> rowInds;
+  external "C" serializeJ(name, numCols, nnz, colPtrs, rowInds) annotation(Include="
```

Comment:
Is it the best approach to add C code to MetaModelica files? Why not add it to the runtime and use it from here? Makes finding errors in the C code way easier I guess.

Reply:
You are probably right. But since this is not meant to stay forever, we didn't care too much.

Reply:
It is still here...